### ISSUE TITLE
Make `get_header` generic over response & bid evaluation function

### ISSUE THREAD

**Issue by mempirate** [Commit-Boost/commit-boost-client#105]
## Context
In the [`get_header_with_proofs`](https://chainbound.github.io/bolt-docs/api/builder#get_header_with_proofs) endpoint of the constraints-API, we reuse practically all the same functionality of the `get_header` call defined in commit-boost PBS. The only differences are:
- A different response (`GetHeaderWithProofsResponse`)
- A bid evaluation function (the proofs need to be valid in order for the bid to be accepted)

It would be great if we could reuse the same `get_header` implementation with all of the existing configuration, without having to copy over all of that functionality. For that, we'd need to make `mev_boost::get_header` generic over the response type and provide an optional generic `eval_bid` closure. Something like the following function signature should make it clear:
```rs
/// Implements https://ethereum.github.io/builder-specs/#/Builder/getHeader
/// Returns 200 if at least one relay returns 200, else 204
pub async fn get_header<S, R, F>(
    params: GetHeaderParams,
    req_headers: HeaderMap,
    state: PbsState<S>,
    eval_bid: Option<F>,
) -> eyre::Result<Option<R>>
where
    S: BuilderApiState,
    R: Serialize + Deserialize<'static>,
    F: Fn(&PbsState<S>, &R) -> bool,
```

Would this be something we could implement? Happy to open a PR if so.

**Comment by mempirate** [comment]
Yes exactly, it will be applied to each incoming block from each relay before selecting the best bid

**Comment by fbrv** [comment]
@mempirate is it still relevant?
